File: src/sc_neurocore/neurons/models/escape_rate.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Gerstner 2000 — stochastic threshold (escape noise model)

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np
import numpy.typing as npt

from sc_neurocore.neurons._stochastic_threshold import (
    DEFAULT_LFSR16_SEED,
    Lfsr16Threshold,
)
from sc_neurocore.utils.numerics import safe_exp


@dataclass
class EscapeRateNeuron:
    """Gerstner 2000 — stochastic threshold (escape noise model).

    Membrane dynamics use the exact constant-current RC flow before evaluating
    the finite-step escape hazard.

    Reference: Gerstner, W. (2000). Neural Comput. 12:43–89.
    """

    v: float = -70.0
    v_rest: float = -70.0
    v_reset: float = -70.0
    v_threshold: float = -50.0
    tau_m: float = 10.0
    rho_0: float = 0.001
    delta_u: float = 3.0
    resistance: float = 1.0
    dt: float = 1.0
    seed: int | None = DEFAULT_LFSR16_SEED

    def __post_init__(self) -> None:
        self._validate_runtime_state()
        self._rng = Lfsr16Threshold(self.seed)

    @property
    def initial_seed(self) -> int:
        """Return the concrete seed used by this instance."""
        return self._rng.initial_seed

    @property
    def rng_state(self) -> int:
        """Return the current canonical LFSR16 state."""
        return self._rng.state

    def _validate_runtime_state(self) -> None:
        for field in ("v", "v_rest", "v_reset", "v_threshold"):
            if not math.isfinite(getattr(self, field)):
                raise ValueError(f"{field} must be finite")
        for field in ("tau_m", "rho_0", "delta_u", "resistance", "dt"):
            value = getattr(self, field)
            if not math.isfinite(value) or value <= 0.0:
                raise ValueError(f"{field} must be finite and positive")

    def _spike_probability(self, voltage: float) -> float:
        if not math.isfinite(voltage):
            raise ValueError("voltage candidate must be finite")
        rate = self.rho_0 * safe_exp((voltage - self.v_threshold) / self.delta_u)
        hazard = rate * self.dt
        if not math.isfinite(hazard) or hazard < 0.0:
            raise ValueError("escape hazard must be finite and non-negative")
        probability = -math.expm1(-hazard)
        if not math.isfinite(probability) or not 0.0 <= probability <= 1.0:
            raise ValueError("spike probability must remain finite and bounded")
        return probability

    def step(self, current: float) -> int:
        if not math.isfinite(current):
            raise ValueError("current must be finite")
        self._validate_runtime_state()

        voltage = self._exact_voltage_candidate(current)
        p_spike = self._spike_probability(voltage)
        if self._rng.trial(p_spike):
            self.v = self.v_reset
            return 1
        self.v = voltage
        return 0

    def simulate(
        self,
        n_steps: int,
        current: float = 0.0,
        backend: str = "auto",
    ) -> tuple[npt.NDArray[np.float64], int]:
        """Run a constant-current trace on Python or one real native backend.

        Every backend receives the complete physical state and the live LFSR
        state. A successful batch atomically commits its final voltage and RNG
        state; unavailable or rejected native runs leave the instance unchanged.
        """
        if isinstance(n_steps, bool) or not isinstance(n_steps, int) or n_steps < 0:
            raise ValueError("n_steps must be a non-negative integer")
        if backend not in ("auto", "python", "rust", "julia", "go", "mojo"):
            raise ValueError(f"backend must be auto/python/rust/julia/go/mojo, got {backend!r}")
        current = float(current)
        if not math.isfinite(current):
            raise ValueError("current must be finite")
        self._validate_runtime_state()

        from sc_neurocore.accel import escape_rate as backends

        selected = backend
        if selected == "auto":
            if backends._HAS_RUST:
                selected = "rust"
            elif backends.ensure_mojo_loaded():
                selected = "mojo"
            elif backends.ensure_go_loaded():
                selected = "go"
            elif backends.ensure_julia_loaded():
                selected = "julia"
            else:
                selected = "python"

        previous_v = self.v
        previous_rng = self.rng_state
        try:
            if selected == "python":
                result = self._simulate_python(n_steps, current)
            else:
                loader = {
                    "rust": lambda: backends._HAS_RUST,
                    "julia": backends.ensure_julia_loaded,
                    "go": backends.ensure_go_loaded,
                    "mojo": backends.ensure_mojo_loaded,
                }[selected]
                if not loader():
                    raise RuntimeError(f"{selected.title()} EscapeRate backend is unavailable.")
                runner = {
                    "rust": backends.simulate_rust,
                    "julia": backends.simulate_julia,
                    "go": backends.simulate_go,
                    "mojo": backends.simulate_mojo,
                }[selected]
                result = runner(
                    self.v,
                    self.v_rest,
                    self.v_reset,
                    self.v_threshold,
                    self.tau_m,
                    self.rho_0,
                    self.delta_u,
                    self.resistance,
                    self.dt,
                    self.rng_state,
                    n_steps,
                    current,
                )
            trace, events, final_v, final_rng = backends._normalise_result(*result)
            trace_array = np.ascontiguousarray(trace, dtype=np.float64)
            spike_count = int(np.sum(events, dtype=np.int64))
            self._rng.restore(final_rng)
            self.v = final_v
        except Exception:
            self.v = previous_v
            self._rng.restore(previous_rng)
            raise

        return trace_array, spike_count

    def _simulate_python(
        self, n_steps: int, current: float
    ) -> tuple[npt.NDArray[np.float64], npt.NDArray[np.uint8], float, int]:
        """Execute the canonical Python recurrence without a surrogate path."""
        trace = np.empty(n_steps, dtype=np.float64)
        events = np.empty(n_steps, dtype=np.uint8)
        for index in range(n_steps):
            events[index] = self.step(current)
            trace[index] = self.v
        return trace, events, self.v, self.rng_state

    def reset(self) -> None:
        self.v = self.v_rest
        self._rng.reset()

    def _exact_voltage_candidate(self, current: float) -> float:
        steady_state = self.v_rest + self.resistance * current
        decay = math.exp(-self.dt / self.tau_m)
        voltage = steady_state + (self.v - steady_state) * decay
        if (
            not math.isfinite(steady_state)
            or not math.isfinite(decay)
            or not math.isfinite(voltage)
        ):
            raise ValueError("voltage candidate must be finite")
        return voltage

File: src/sc_neurocore/accel/rust/safety/escape_rate.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for escape_rate

#[derive(Debug, Clone)]
pub struct EscapeRateNeuron {
    pub v: f64,
    pub v_rest: f64,
    pub v_reset: f64,
    pub v_threshold: f64,
    pub tau_m: f64,
    pub rho_0: f64,
    pub delta_u: f64,
    pub resistance: f64,
    pub dt: f64,
    pub rng_state: u16,
    pub initial_seed: u16,
}

impl EscapeRateNeuron {
    pub fn new() -> Self {
        Self::new_with_seed(0xACE1)
    }

    pub fn new_with_seed(seed: u16) -> Self {
        let initial_seed = if seed == 0 { 0xACE1 } else { seed };
        Self {
            v: -70.0_f64,
            v_rest: -70.0_f64,
            v_reset: -70.0_f64,
            v_threshold: -50.0_f64,
            tau_m: 10.0_f64,
            rho_0: 0.001_f64,
            delta_u: 3.0_f64,
            resistance: 1.0_f64,
            dt: 1.0_f64,
            rng_state: initial_seed,
            initial_seed,
        }
    }

    pub fn step(&mut self, i_ext: f64) -> Result<i32, &'static str> {
        if !i_ext.is_finite() {
            return Err("escape rate input current must be finite");
        }
        if !validate_escape_rate(self) {
            return Err("escape rate state parameters must be finite and positive");
        }

        let v_inf = self.v_rest + self.resistance * i_ext;
        let decay = (-self.dt / self.tau_m).exp();
        let next_v = v_inf + (self.v - v_inf) * decay;
        if !v_inf.is_finite() || !decay.is_finite() || !next_v.is_finite() {
            return Err("escape rate membrane candidate must remain finite");
        }
        let hazard = self.rho_0 * safe_exp((next_v - self.v_threshold) / self.delta_u) * self.dt;
        if !hazard.is_finite() || hazard < 0.0 {
            return Err("escape rate hazard must remain finite and non-negative");
        }
        let p_spike = -(-hazard).exp_m1();
        if !p_spike.is_finite() || !(0.0..=1.0).contains(&p_spike) {
            return Err("escape rate spike probability must remain finite and bounded");
        }
        let mut sample = self.rng_state;
        for _ in 0..8 {
            let feedback = ((sample >> 0) ^ (sample >> 2) ^ (sample >> 3) ^ (sample >> 5)) & 1;
            sample = (sample >> 1) | (feedback << 15);
        }
        let threshold = if p_spike <= 0.0 {
            0_u32
        } else if p_spike >= 1.0 {
            65_536_u32
        } else {
            (p_spike * 65_535.0).floor() as u32 + 1
        };
        self.rng_state = sample;
        if u32::from(sample) < threshold {
            self.v = self.v_reset;
            return Ok(1);
        }
        self.v = next_v;
        Ok(0)
    }

    pub fn reset(&mut self) {
        self.v = self.v_rest;
        self.rng_state = self.initial_seed;
    }
}

pub fn validate_escape_rate(state: &EscapeRateNeuron) -> bool {
    state.v.is_finite()
        && state.v_rest.is_finite()
        && state.v_reset.is_finite()
        && state.v_threshold.is_finite()
        && state.tau_m.is_finite()
        && state.tau_m > 0.0
        && state.rho_0.is_finite()
        && state.rho_0 > 0.0
        && state.delta_u.is_finite()
        && state.delta_u > 0.0
        && state.resistance.is_finite()
        && state.resistance > 0.0
        && state.dt.is_finite()
        && state.dt > 0.0
        && state.rng_state != 0
}

fn safe_exp(x: f64) -> f64 {
    x.clamp(-700.0, 700.0).exp()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_escape_rate_new() {
        let state = EscapeRateNeuron::new();
        assert!(state.v.is_finite());
        assert!(validate_escape_rate(&state));
    }

    #[test]
    fn test_escape_rate_step() {
        let mut state = EscapeRateNeuron::new_with_seed(42);
        let spike = state.step(10.0).expect("valid step must succeed");
        assert!(spike == 0 || spike == 1);
    }

    #[test]
    fn test_seeded_sequence_and_reset_are_replayable() {
        let mut state = EscapeRateNeuron::new_with_seed(42);
        let first: Vec<i32> = (0..1000)
            .map(|_| state.step(30.0).expect("valid step"))
            .collect();
        let final_rng = state.rng_state;
        state.reset();
        let replay: Vec<i32> = (0..1000)
            .map(|_| state.step(30.0).expect("valid step"))
            .collect();
        assert_eq!(first, replay);
        assert_eq!(state.rng_state, final_rng);
    }

    #[test]
    fn test_exact_flow_matches_closed_form() {
        let mut state = EscapeRateNeuron::new();
        state.v = -65.0;
        state.dt = 5.0;
        state.rho_0 = 1.0e-12;
        let current = 10.0;
        let v0 = state.v;
        let v_inf = state.v_rest + state.resistance * current;
        let euler =
            v0 + (-(v0 - state.v_rest) + state.resistance * current) / state.tau_m * state.dt;
        let expected = v_inf + (v0 - v_inf) * (-state.dt / state.tau_m).exp();

        let spike = state.step(current).expect("valid step must succeed");
        assert_eq!(spike, 0);
        assert!((state.v - expected).abs() < 1e-14);
        assert!((state.v - euler).abs() > 1e-3);
    }

    #[test]
    fn test_invalid_runtime_state_does_not_mutate_voltage() {
        let mut state = EscapeRateNeuron::new();
        state.v = -65.0;
        state.delta_u = 0.0;

        assert!(state.step(1.0).is_err());
        assert_eq!(state.v, -65.0);
    }

    #[test]
    fn test_non_finite_update_does_not_mutate_voltage() {
        let mut state = EscapeRateNeuron::new();
        state.v = -65.0;
        state.v_threshold = 1.0e308;
        state.resistance = 1.0e308;

        assert!(state.step(1.0e308).is_err());
        assert_eq!(state.v, -65.0);
    }

    #[test]
    fn test_non_finite_hazard_does_not_mutate_voltage() {
        let mut state = EscapeRateNeuron::new();
        state.v = -50.0;
        state.rho_0 = 1.0e308;
        state.dt = 1.0e308;

        assert!(state.step(0.0).is_err());
        assert_eq!(state.v, -50.0);
    }
}
